sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -542,29 +542,42 @@ object PushFoldableIntoBranches extends Rule[LogicalPlan] with PredicateHelper {
 
   def apply(plan: LogicalPlan): LogicalPlan = plan transform {
     case q: LogicalPlan => q transformExpressionsUp {
+      case a: Alias => a // Skip an alias.
```

Comment:
`Alias` is not the only exception, we can't apply this optimization for `Generator` as well, as the logical plan `Generate` requires explicit type `Generator`.

It happened many times that an optimization rule introduces bugs because it uses denylist instead of allowlist. Let's avoid similar mistakes here, and explicitly list what expressions we should support.

An initial list from my mind:
1. IsNull, IsNotNull
2. UnaryMathExpression
3. String2StringExpression
4. Cast
5. BinaryComparison
6. BinaryArithmetic

Reply:
Do you think `BinaryExpression` also has this issue?

Reply:
I'm not sure, but it's safer to start with an allowlist.  We can extend it later.

Reply:
https://github.com/apache/spark/pull/30955